### PR TITLE
feature/V1-078 - endpoint for pending courses

### PR DESCRIPTION
**Pull request by zhuk-vezunchikz**
- created endpoint for pending courses

### CODE REVIEW

File: backend/src/db/providers/clientCourseProvider.ts [hunk 1]
```diff
@@ -22,9 +24,13 @@ const applyCourseProvider = async (courseId: string, userId: string, progressDto
...
     user: userId,
     course: courseId,
-    status: CourseStatus.approved,
+    status: CourseStatus.pending,
     progress: progressDto,
   });
+  await UserModel.updateMany(
+    { role: USER_ROLES.MANAGER },
+    { $push: { pendingCourses: applyedCourse } },
+  );
```
Comment (SaXaPhonist) [inline backend/src/db/providers/clientCourseProvider.ts:33]: i think we need to pass additional agr ManagerId to update only manager we need not all as in this case
cause here all manager will get this new applyed course

File: backend/src/db/providers/userProvider.ts [hunk 2]
```diff
@@ -1,7 +1,24 @@
...
+          model: 'Courses',
+          select: 'title',
+        },
+        {
+          path: 'user',
+          model: 'User',
+          select: 'firstName lastName position avatar',
+        },
+      ],
+    })
```
Comment (SaXaPhonist) [inline backend/src/db/providers/userProvider.ts:20]: we should not  populate our user here, just stay with array of pending-courses id
you can create new  provider and call it for example "getPendingCourseProvider" - to get only pending course

File: backend/src/interfaces/IPendingCoursesList/IPendingCourse.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { ObjectId } from 'mongoose';
+
+interface IPendingCourse {
+  _id?: string;
+  user: ObjectId;
+  course: ObjectId;
+  time: Date;
+  status: ObjectId;
+}
+
```
Comment (SaXaPhonist) [inline backend/src/interfaces/IPendingCoursesList/IPendingCourse.ts:10]: pending course it's just omited type of clientCourses
 you can inherit this type from clientCourse

File: backend/src/interfaces/Ientities/Iusers.ts [hunk 4]
```diff
@@ -14,6 +14,7 @@ interface IUser {
   courses: ObjectId[];
   group: string;
   employees: ObjectId[];
+  pendingCourses: ObjectId[];
```
Comment (SaXaPhonist) [inline backend/src/interfaces/Ientities/Iusers.ts:17]: don't foget that sometimes it's field can containe populated objects, so you need add "or" and use type from prev

**Review by SaXaPhonist — CHANGES_REQUESTED** [review]

File: backend/src/routing/pendingCourses.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { Router } from 'express';
```
Comment (dapodev) [inline backend/src/routing/pendingCourses.ts:1]: Empty line to separate lib imports.

File: backend/src/routing/pendingCourses.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+import { Params } from '../enums/routesEnum';
+import withAuth from '../middlewares/authMiddleware';
+import { USER_ROLES } from '../config/constants';
+import getPendingCourses from '../controllers/pendingCourses/getPendingCourses';
```
Comment (dapodev) [inline backend/src/routing/pendingCourses.ts:5]: All this stuff should be imported using absolute paths! (you're importing components from other src-level dir)

File: backend/src/db/providers/clientCourseProvider.ts [hunk 7]
```diff
@@ -2,6 +2,8 @@ import { IProgress } from 'interfaces/ICourses/IQueryCourses';
 import CourseStatus from 'enums/coursesEnums';
 
 import ClientCourseModel from '../models/ClientCourses';
+import UserModel from '../models/User';
+import { USER_ROLES } from '../../config/constants';
```
Comment (dapodev) [inline backend/src/db/providers/clientCourseProvider.ts:6]: This import should be absolute!

**Review by dapodev — CHANGES_REQUESTED** [review]
Just some refactor is needed. 🐤

File: backend/src/controllers/pendingCourses/getPendingCourses.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import { TMiddlewareCall } from 'interfaces/commonMiddleware';
+import { IUser } from 'interfaces/Ientities/Iusers';
+import { getUserProvider } from 'db/providers/userProvider';
+import { verifyAccessToken } from 'utils/auth/authUtils';
+
+const getPendingCourses = async (req: Request, res: Response, next: TMiddlewareCall) => {
+  try {
+    const accessToken = req.headers.authorization?.split(' ')[1];
+    if (accessToken) {
+      const decodetToken = await verifyAccessToken(accessToken);
```
[the remark this answers is not in the document]
Comment (zhuk-vezunchikz) [reply backend/src/controllers/pendingCourses/getPendingCourses.ts:13]: done

File: backend/src/db/models/PendingCoursesList.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  course: { type: Schema.Types.ObjectId, ref: 'Courses' },
+  status: { type: Schema.Types.ObjectId, ref: 'clientCourses' },
+  time: { type: Date },
+});
+
+const PendingCoursesListModel = model<IPendingCourse>(
+  'pendingCourseSchema',
+  pendingCourseSchema,
+  'pendingCourseSchema',
+);
```
[the remark this answers is not in the document]
Comment (zhuk-vezunchikz) [reply backend/src/db/models/PendingCoursesList.ts:16]: done

File: backend/src/db/providers/clientCourseProvider.ts (hunk 1 above)
Comment (zhuk-vezunchikz) [reply backend/src/db/providers/clientCourseProvider.ts:33]: done

File: backend/src/db/providers/userProvider.ts (hunk 2 above)
Comment (zhuk-vezunchikz) [reply backend/src/db/providers/userProvider.ts:20]: done

File: backend/src/db/providers/clientCourseProvider.ts (hunk 7 above)
Comment (zhuk-vezunchikz) [reply backend/src/db/providers/clientCourseProvider.ts:6]: done

File: backend/src/controllers/pendingCourses/getPendingCourses.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+import { isError } from 'utils/typeGuards/isError';
+import { TMiddlewareCall } from 'interfaces/commonMiddleware';
+import { IUser } from 'interfaces/Ientities/Iusers';
+import { getPendingCoursesProvider } from 'db/providers/pendingCoursesProvider';
+
+const getPendingCourses = async (req: Request, res: Response, next: TMiddlewareCall) => {
+  const { id: userId } = res.locals;
+  try {
+    const user: IUser = await getPendingCoursesProvider(userId);
```
Comment (dapodev) [inline backend/src/controllers/pendingCourses/getPendingCourses.ts:11]: You should better choose which way you will use typing: explicitly or implicitly. The thing is that you wrote the same pieces of code in different ways in different files. 
image
Let's just not use explicit typing where we can really avoid it.

File: backend/src/db/providers/clientCourseProvider.ts [hunk 11]
```diff
@@ -1,7 +1,9 @@
+import { ObjectId } from 'mongoose';
```
Comment (dapodev) [inline backend/src/db/providers/clientCourseProvider.ts:1]: Empty line! Mongoose is a library and it should stand separately.

File: backend/src/db/providers/pendingCoursesProvider.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
+import UserModel from 'db/models/User';
```
Comment (dapodev) [inline backend/src/db/providers/pendingCoursesProvider.ts:1]: Here is the exact case when you should use relative import. Just have a look: this file positioned at the same src-level folder as UserModel (`db`).

**Review by dapodev — DISMISSED** [review]
Well, I'll approve that, but remember to do these changes before to merge PR. 🙏🏻

File: backend/src/db/providers/pendingCoursesProvider.ts (hunk 12 above)
Comment (zhuk-vezunchikz) [reply backend/src/db/providers/pendingCoursesProvider.ts:1]: done

File: backend/src/controllers/pendingCourses/getPendingCourses.ts (hunk 10 above)
Comment (zhuk-vezunchikz) [reply backend/src/controllers/pendingCourses/getPendingCourses.ts:11]: done

File: backend/src/db/providers/clientCourseProvider.ts (hunk 11 above)
Comment (zhuk-vezunchikz) [reply backend/src/db/providers/clientCourseProvider.ts:1]: done

File: backend/src/controllers/clientCourses/applyCourse.ts [hunk 13]
```diff
@@ -7,6 +7,9 @@ import { generateProgressDto } from 'utils/dto/dtoUtils';
 import { INITIAL_INDX } from 'config/constants';
 import { materialsCounterProvider } from 'db/providers/courseProvider';
 import { checkCourseDuplicates } from 'utils/validation/checkDuplicates';
+import { getPendingCoursesProvider } from 'db/providers/pendingCoursesProvider';
+import { IUser } from '../../interfaces/Ientities/Iusers';
+import { IClientCourse } from '../../interfaces/Ientities/IclientCourses';
```
Comment (k-krasilnikova) [inline backend/src/controllers/clientCourses/applyCourse.ts:12]: Absolute imports!!

File: backend/src/db/providers/clientCourseProvider.ts [hunk 14]
```diff
@@ -1,7 +1,9 @@
...
+
 import { IProgress } from 'interfaces/ICourses/IQueryCourses';
 import CourseStatus from 'enums/coursesEnums';
-
 import ClientCourseModel from '../models/ClientCourses';
+import UserModel from '../models/User';
```
Comment (k-krasilnikova) [inline backend/src/db/providers/clientCourseProvider.ts:6]: !

File: backend/src/db/providers/pendingCoursesProvider.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
+import UserModel from '../models/User';
```
Comment (k-krasilnikova) [inline backend/src/db/providers/pendingCoursesProvider.ts:1]: !

**Review by k-krasilnikova — CHANGES_REQUESTED** [review]

**Review by SaXaPhonist — APPROVED** [review]

**Review by dapodev — APPROVED** [review]

**Review by k-krasilnikova — APPROVED** [review]

**Pull request merged**